Add per-tenant provisioning throttling

dbwiddis opened the pull request:
### Description

Adds settings to control per-tenant limits on simultaneous provision, reprovision, and deprovision.

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [x] Commits are signed per the DCO using `--signoff`.
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/flow-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dbwiddis commented:
tests failing because of jar hell caused by https://github.com/opensearch-project/ml-commons/pull/3475

owaiskazi19 commented on src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java:
```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+            return; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.get(tenantId);
+        if (count != null) {
+            count.decrementAndGet();
+        }
```
Don't we have to remove tenantId from the map here?
```suggestion
        if (count != null && count.decrementAndGet() <= 0) {
        executionsMap.remove(tenantId);
    }
```

dbwiddis replied:
The code logic would work the same here.  But I can see how filling up a static map with a bunch of 0's could leave a bunch of zero map entries.

dbwiddis replied:
Your code suggestion still leaves a race condition between the decrement and get, and the removal.

I think we could use computeIfPresent, though, and it removes the entry on null.

amitgalitz replied:
+1 on computeIfPresent so if count is 0 then we should remove entry from map

amitgalitz replied:
Additionally could a race condition here cause a negative count?

dbwiddis replied:
> could a race condition here cause a negative count?

I don't think so?  The code to decrement is only run after there was an increment.

owaiskazi19 commented on src/main/java/org/opensearch/flowframework/common/FlowFrameworkSettings.java:
```diff
@@ -129,6 +133,25 @@ public class FlowFrameworkSettings {
+    /** This setting sets max workflows that can be simultaneously provisioned, or reprovisioned by the same tenant */
+    public static final Setting<Integer> MAX_ACTIVE_PROVISIONS_PER_TENANT = Setting.intSetting(
+        "plugins.flow_framework.max_active_provisions_per_tenant",
+        2,
```
Why are we keeping the default value as 2 here?

dbwiddis replied:
I want the default value to be less than our thread pool size. 

What other default value do you suggest?

owaiskazi19 commented on src/main/java/org/opensearch/flowframework/transport/ReprovisionWorkflowTransportAction.java:
```diff
@@ -452,8 +460,15 @@ private void executeWorkflow(
-                }, exception -> { logger.error("Failed to update workflow state for workflow {}", workflowId, exception); })
+                }, exception -> {
+                    // We've completed provisioning asynchronously but failed state update
+                    TenantAwareHelper.releaseProvision(tenantId);
```
In the concurrent map should we also store the status of provisioning along with the tenantId? To keep a track for which tenantId we had a successful provisioning

dbwiddis replied:
Not sure what you're asking. This is intended to be a very lightweight throttling mechanism, not a stats collector.

owaiskazi19 commented on src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java:
```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+                    delegate.onFailure(e);
+                }
+            }
+        });
```
Can use runAfter to reduce the method completely
```java
return ActionListener.notifyOnce(ActionListener.runAfter(delegate, () -> releaseDeprovision(tenantId)));
```

dbwiddis replied:
In theory, yes, we could use `runBefore()` or `runAfter()` here because we are calling `releaseDeprovision()` in both cases.  In fact, the code here is exactly what happens on `runBefore()`.  

However, this is just a lucky break because we use synchronous code for deprovisioning, so our action listener handling is simpler.  We can _not_ do it for the provision and reprovision code because we only call `releaseProvision()` on failure and _not_ on success.

In which case I'd prefer to keep this method implementation similar to the adjacent method in the same class.

owaiskazi19 commented on src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java:
```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+                    delegate.onFailure(e);
+                }
+            }
+        });
```
Same as below

dbwiddis replied:
We only release on failure here so we can't wrap it with the same method in both cases.

owaiskazi19 commented on src/test/java/org/opensearch/flowframework/rest/RestWorkflowProvisionTenantAwareIT.java:
```diff
@@ -0,0 +1,376 @@
+        response = makeRequest(otherTenantRequest, DELETE, WORKFLOW_PATH + otherWorkflowId);
+        assertOK(response);
+        map = responseToMap(response);
+        assertEquals(otherWorkflowId, map.get(DOC_ID).toString());
```
We are doing a bunch of operations in this test case. We could split the tests with `whenThrottlingEnabled`, `whenMultitenantIsEnabled`, `whenThrottlingisDisabled` and  `whenMultitenantIsDisabled`

dbwiddis replied:
I used the same pattern as my other ITs, which rely on a long sequence of operations, most of which are the same and only some of which are different.   Splitting them up would increase duplication.

owaiskazi19 reviewed (commented):
Looks good overall with comments

dbwiddis commented on src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java:
```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+        if (tenantId == null) {
+            return true; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.computeIfAbsent(tenantId, k -> new AtomicInteger(0));
+        if (count.incrementAndGet() <= maxExecutions) {
+            return true;
+        } else {
+            count.decrementAndGet(); // Rollback the increment
+            return false;
+        }
```
> We could use `compareAndSet` to avoid any race condition due to increment and decrement

I considered almost exactly your suggestion but there's still a race condition between your `.get()` on line 171 and the comparison on line 173.  In which case `compareAndSet()` might fail even if the change in `count` was a decrement or an increment below maxExecutions.

amitgalitz commented on src/main/java/org/opensearch/flowframework/common/FlowFrameworkSettings.java:
```diff
@@ -129,6 +133,25 @@ public class FlowFrameworkSettings {
+    /** This setting sets max workflows that can be simultaneously deprovisioned by the same tenant */
+    public static final Setting<Integer> MAX_ACTIVE_DEPROVISIONS_PER_TENANT = Setting.intSetting(
+        "plugins.flow_framework.max_active_deprovisions_per_tenant",
+        1,
```
I think we should keep it the same as provision and at 2 at least

amitgalitz commented on src/main/java/org/opensearch/flowframework/common/FlowFrameworkSettings.java:
```diff
@@ -129,6 +133,25 @@ public class FlowFrameworkSettings {
+        "plugins.flow_framework.max_active_provisions_per_tenant",
+        2,
+        1,
+        4,
```
Should we also make our threadpool size dynamic and the max higher here incase we have larger nodes that can accommodate that?

dbwiddis replied:
Yes.

dbwiddis commented:
@owaiskazi19 @amitgalitz Thanks for your review!  

I've addressed some comments in https://github.com/opensearch-project/flow-framework/pull/1074/commits/6fbf71842500eb02680e92dc4e2ba5b2bca756f6:
 - used `computeIfPresent()` to remove from the map when it hits 0
 - used `ActionListener.runBefore()` in the deprovision case; it is a lot simpler

I'd like to handle changing the value of settings/limits and defaults in another PR (followed by a discussion issue).

I'll consider refactoring all of the multitenant ITs at a later date.

Other comments I've responded that I don't think changes are needed; I've added some comments in code explaining why.  In particular I don't think there's a easy way to replace the increment/rollback logic with compareAndSet without really overcomplicating things.

owaiskazi19 reviewed (approved):
Thanks for addressing the comments
